### PR TITLE
Fix one-time kicking in simulation

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

This PR fixes how the robots kick and chip in simulation. They are now able to properly perform one-time shots.

* The ball is now heavily damped when it hits a robot's dribbler
* Some momentum is conserved and put back into the kick. This helps approximate how if we kick a ball moving towards us, it will leave with an even higher velocity.
* The kicking logic has mostly been moved up to the `SimulatorRobot` rather than living in the `PhysicsBall` class, which is both a bit more proper and was easier to make this work.

### Testing Done

Updated existing unit tests

### Resolved Issues

None, this was an issue discovered in #1545 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/firmware/shared/util.c [hunk 1]
```diff
@@ -5,7 +5,7 @@
 float fmax_of_array(float array[], unsigned size)
 {
     unsigned i;
-    float max_value;
+    float max_value = 0.0f;
```
Comment (MathewMacDougall) [inline src/firmware/shared/util.c:8]: Note: I built with optimizations and compilation failed because this value is "possibly uninitialized", which is true if the `size=0`. That's why these changes exist (not sure why this wasn't caught with a normal fastbuild).

File: src/software/simulation/physics/physics_ball_test.cpp [hunk 2]
```diff
@@ -407,23 +387,27 @@ TEST(PhysicsBallTest, test_chip_ball_with_collisions)
     obstacle_body->CreateFixture(&obstacle_fixture_def);
 
     EXPECT_FALSE(physics_ball.isInFlight());
-    physics_ball.chip(Vector(2, 0));
+    physics_ball.setInFlightForDistance(2.0f);
     EXPECT_TRUE(physics_ball.isInFlight());
 
-    for (unsigned int i = 0; i < 100; i++)
+    for (unsigned int i = 0; i < 300; i++)
```
Comment (kelvinkoon) [inline src/software/simulation/physics/physics_ball_test.cpp:404]: What's the reason for changing this?

File: src/software/simulation/simulator_robot.cpp [hunk 3]
```diff
@@ -133,10 +147,28 @@ void SimulatorRobot::chip(float distance_m)
...
+            // until it has travelled the desired chip distance.
+            ball->setInFlightForDistance(distance_m);
+
+            // Assume the ball is chipped at a 45 degree angle
+            // TODO: Use a robot-specific constant
+            // https://github.com/UBC-Thunderbots/Software/issues/1179
+            Angle chip_angle = Angle::fromDegrees(45);
+            // Use the formula for the Range of a parabolic projectile
+            // Rearrange to solve for the initial velocity
+            // See
```
Comment (kelvinkoon) [inline src/software/simulation/simulator_robot.cpp]: This "see" comment isn't necessary.

File: src/software/simulation/simulator_robot.h [hunk 4]
```diff
@@ -299,4 +299,18 @@ class SimulatorRobot
...
+    // this behavior to match real life. These values have been manually tuned
+    // such that the robots are able to hit one-time shots in simulation with
+    // sufficient accuracy.
+    static constexpr double DRIBBLER_HEAD_ON_DAMPING       = 0.7;
+    static constexpr double DRIBBLER_PERPENDICULAR_DAMPING = 0.61;
+    // A value in the range [0, 1] that indicates how much momentum is conserved when the
+    // ball is kicked. Higher values will cause the ball to be kicked with an even greater
+    // velocity if it had an initial non-zero velocity when being kicked.
+    // This value is a very rough estimate of real-world behaviour, so that the ball
+    // will be kicked slightly faster it it's kicked with some initial velocity.
```
Comment (jonathanlew) [inline src/software/simulation/simulator_robot.h]:
```suggestion
    // will be kicked slightly faster it it entered the kicker with some initial velocity.
```

File: src/software/simulation/simulator_robot.cpp [hunk 5]
```diff
@@ -349,6 +375,20 @@ void SimulatorRobot::onDribblerBallContact(PhysicsRobot *physics_robot,
 void SimulatorRobot::onDribblerBallStartContact(PhysicsRobot *physics_robot,
                                                 PhysicsBall *physics_ball)
 {
+    // Damp the ball when it collides with the dribbler. We damp each component
+    // of the ball's momentum separately so we have the flexibility to tune this
+    // behavior to match real life.
+    Vector ball_momentum       = physics_ball->momentum();
+    Vector robot_facing_vector = Vector::createFromAngle(physics_robot->orientation());
+    Vector robot_perp_vector   = robot_facing_vector.perpendicular();
```
Comment (jonathanlew) [inline src/software/simulation/simulator_robot.cpp:405]: does this account for a moving robot?

**Review by jonathanlew — CHANGES_REQUESTED** [review]

File: src/software/simulation/simulator_robot.cpp (hunk 5 above)
Comment (MathewMacDougall) [reply src/software/simulation/simulator_robot.cpp:405]: No, but I think this is a sufficient approximation

**Review by kelvinkoon — COMMENTED** [review]
Quick glance seems good overall.

File: src/software/simulation/simulator_robot.cpp (hunk 5 above)
Comment (jonathanlew) [reply src/software/simulation/simulator_robot.cpp:405]: Damping should use the relative velocity between the ball and the robot. (Arguably, perfect damping is a sufficient approximation, so if we're going to do damping, we should do it right)

**Review by jonathanlew — CHANGES_REQUESTED** [review]

File: src/software/simulation/simulator_robot.cpp (hunk 5 above)
Comment (MathewMacDougall) [reply src/software/simulation/simulator_robot.cpp:405]: Perfect damping isn't a perfect approximation (at least to real-life). If it was perfectly damped, by the time the chicker activates there'd be no remaining momentum to "conserve", so we'd always kick at exactly the desired speed. That would be ideal, but not "realistic" which is why we have this medium damping.

Comment (MathewMacDougall) [reply src/software/simulation/simulator_robot.cpp:405]: Will add robot velocity, standby

File: src/software/simulation/physics/physics_ball.cpp [hunk 6]
```diff
@@ -121,25 +104,31 @@ bool PhysicsBall::isTouchingOtherObject() const
...
-    bool chip_in_progress = chip_origin.has_value();
-    if (chip_in_progress)
+    bool ball_currently_in_flight = in_flight_origin.has_value();
+    if (ball_currently_in_flight)
     {
-        double current_chip_distance_meters =
-            (getBallState().position() - chip_origin.value()).length();
+        double current_in_flight_distance_meters =
+            (position() - in_flight_origin.value()).length();
         // Once the ball is in flight, is can only stop being in flight once it has
```
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/software/simulation/physics/physics_ball.cpp]: fixed

File: src/software/simulation/physics/physics_ball_test.cpp (hunk 2 above)
Comment (MathewMacDougall) [reply src/software/simulation/physics/physics_ball_test.cpp:404]: Ball is now moving at 1m/s, and we want to make sure it passes all the way through the object at (2, 0), so this just makes sure there's enough time for that to happen

File: src/software/simulation/simulator_robot.cpp (hunk 3 above)
Comment (MathewMacDougall) [reply src/software/simulation/simulator_robot.cpp]: fixed

File: src/software/simulation/simulator_robot.cpp (hunk 5 above)
Comment (MathewMacDougall) [reply src/software/simulation/simulator_robot.cpp:405]: Added robot velocity as well and added tests. I agree this is better.

**Review by garethellis0 — DISMISSED** [review]

**Review by jonathanlew — DISMISSED** [review]
LGTM! Make sure to post a video on slack

**Review by akhilveeraghanta — DISMISSED** [review]

**Comment by MathewMacDougall** [comment]
Fixed build, need re-approval

**Review by jonathanlew — APPROVED** [review]

**Review by kelvinkoon — APPROVED** [review]
:footprints:

**Review by garethellis0 — APPROVED** [review]
